### ISSUE TITLE
Monitoring all the things

### ISSUE THREAD

**Issue by sarayourfriend** [WordPress/openverse-api#391]
## Problem

We currently have very little visibility into the health of our individual endpoints.

## Description

Using Grafana and statsd we should be able to gather some relevant metrics that we can monitor (and potentially work on improving, if need be!)

We should look into gathering at least the following for each endpoint:

1. Timings per method
2. Raw request count per method
3. Status codes per method

A potential template for these could be something like:

- `view-timings.{view_name}.{request_method}` -> time series
- `view-frequency.{view_name}.{request_method}` -> simple incremented counter
- `view-status.{view_name}.{request_method}.{response_status_code}` -> simple incremented counter

We'll need to invest some time setting up anomaly detection. There's a few examples of how to do this online using Grafana/Graphite:

http://www.chrisrimondi.com/2015/07/anomaly-detection-with-timeshift-in.html

To be honest this gets into some pretty mathy stuff that I'm not really all too good with! I'm hoping someone else will be able to help sort out exactly what we'll need to do with all of graphites available functions to be able to monitor anomalies in grafana.

Once we've figured out how to do anomaly detection, we can set up some dashboards in Grafana for it!

We'll also want to make sure we can configure Grafana via terraform so that our dashboards aren't _just_ stored in the Grafana database. Luckily there is a terraform provider for Grafana that looks pretty good! https://registry.terraform.io/providers/grafana/grafana/latest/docs/resources/dashboard

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

**Comment by sarayourfriend** [comment]
I've been doing some reading today about Prometheus and comparing it to Graphite and I've come to a few conclusions about Prometheus vs Graphite:

1. Prometheus has a lot more examples and online resources for how to accomplish things like anomaly detection and scaling... generally there's tons of blog posts for Prometheus compared to Graphite which appears to be less widely discussed in the open
2. Prometheus has much better documentation (at least in my experience so far) that seems easier to understand for someone who isn't a big math-head like myself
3. Graphite is a general purpose time-series collection and visualization tool which can be used for anything including system monitoring, whereas Prometheus is specifically a system-monitoring tool. This means Prometheus has clearer handles for system monitoring.
4. Prometheus has the backing of the Linux Foundation, whereas Graphite is a community project—both of these approaches have trade-offs that we know well from our experience with Vue 🙂 

Things that don't matter now but might matter in the future:
1. Prometheus has officially supported client libraries that cover both Python and JavaScript (which Graphite covers as well) in addition to several other popular languages like Java, C++, Go and the .NET family, which Graphite does not support. I don't anticipate us branching out in language choice any time soon but if we ever had to delve into JVM stuff in particular (there's quite a bit of data engineering stuff in Java or especially Scala, I think?) we'd be able to monitor it using the existing infrastructure
2. Prometheus has a much clearer scaling story compared to Graphite. Prometheus has scaling built into it whereas Graphite requires third-party tools to accomplish replication for example. We're not currently at a scale where I think we would benefit from that but looking into the future I sure hope that eventually we do have to worry about it 🙂 

On the other hand, Graphite's more general approach has some advantages. Like I said above Graphite is a general purpose time-series collection and visualization tool. That means we could use it to collect any kind of time-series data, not just system monitoring stuff. This also means it's a lot more "plug-and-play" than the "batteries included" Prometheus. Whereas Prometheus has anomaly detection and alarming built in, Graphite requires a separate library ([Skyline](https://github.com/earthgecko/skyline)) for it. Both can accomplish the same thing in this case, but Graphite requires a little extra work than Prometheus. On the other hand Graphite can easily accomodate collecting business metrics, which Prometheus isn't super appropriate for.

Basically we can think of it as Graphite being a much wider net than Prometheus but that will require more self-service and stitching together to accomplish things that Prometheus, with a narrower scope, is able to do all by itself. This makes Graphite a potentially more powerful tool for us, but would also require more time up front to make it work well for us, whereas Prometheus might get us up and running with alerting and things like that more quickly.

**Comment by obulat** [comment]
I have almost no knowledge in this area, but my vote is for "batteries included" library, because we have a smallish team and it would be nice to have analytics sooner than later.

**Comment by AetherUnbound** [comment]
Thanks for the great explanation! For me, this point is big:
> Prometheus has officially supported client libraries that cover both Python and JavaScript (which Graphite covers as well) in addition to several other popular languages like Java, C++, Go and the .NET family, which Graphite does not support. I don't anticipate us branching out in language choice any time soon but if we ever had to delve into JVM stuff in particular (there's quite a bit of data engineering stuff in Java or especially Scala, I think?) we'd be able to monitor it using the existing infrastructure

I haven't work with the JVM _a ton_, but you're correct that there are some data technologies that include it in their stack. Having good integration with that could be really helpful down the line! I'm also with Olga; we can build our own business reports pretty easily with Airflow, I'd rather have a precision tool with batteries included for this monitoring.

**Comment by rbadillap** [comment]
Thank you for bringing this topic to the table. Monitoring is a very important issue for the number of services that we currently have in our stack.

I support the topic of introducing the best tools to achieve the necessary visibility of what happens in our applications, however, I would introduce this topic slowly so that we can all understand what the true purpose of this task is and how we can all contribute in it.

I'm a fan of Prometheus, not only because of the number of [extensions and integrations it has](https://prometheus.io/docs/instrumenting/exporters/), but because of how clear they are in getting the message across about metrics. They even give us a [great article on the main difference between Graphite and Prometheus themselves](https://prometheus.io/docs/introduction/comparison/).

My advice, Prometheus and Grafana are the winning combination. Not only to obtain metrics from the server ([node_exporter](https://github.com/prometheus/node_exporter) created by Prometheus), or from the containers ([cAdvisor](https://github.com/google/cadvisor) created by Google) but also because we can create our own metrics and expose them in an endpoint ([here an example](https://github.com/brancz/prometheus-example-app)).

**Comment by sarayourfriend** [comment]
Awesome @rbadillap! Thanks for the input and confirmation that Prometheus is probably the best path forward 🎉

**Comment by obulat** [comment]
@sarayourfriend , do you think we can close this issue now that we have an [RFC for monitoring](https://github.com/WordPress/openverse/pull/193)?
